service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessService.java

```diff
@@ -82,19 +90,31 @@ public AzureStorageAccessService(
     this.azureConfiguration = azureConfiguration;
     this.workspaceService = workspaceService;
     this.storageAccountCache = new ConcurrentHashMap<>();
+    this.samUserCache = Collections.synchronizedMap(new PassiveExpiringMap<>(10, TimeUnit.SECONDS));
+    this.storageContainerResourceCache =
+        Collections.synchronizedMap(new PassiveExpiringMap<>(10, TimeUnit.SECONDS));
+    this.storageContainerPermissionsCache =
+        Collections.synchronizedMap(new PassiveExpiringMap<>(10, TimeUnit.SECONDS));
   }
 
   private BlobContainerSasPermission getSasTokenPermissions(
       AuthenticatedUserRequest userRequest,
       UUID storageContainerUuid,
       String samResourceName,
       String desiredPermissions) {
-    final List<String> containerActions =
-        Rethrow.onInterrupted(
-            () ->
-                samService.listResourceActions(
-                    userRequest, samResourceName, storageContainerUuid.toString()),
-            "listResourceActions");
+    List<String> containerActions;
+    var cacheKey = new StorageContainerCacheKey(userRequest.getSubjectId(), storageContainerUuid);
+    if (storageContainerPermissionsCache.containsKey(cacheKey)) {
+      containerActions = storageContainerPermissionsCache.get(cacheKey);
+    } else {
+      containerActions =
```

Comment:
Do you know why this one in particular has the `Rethrow.onInterrupted` handling?

Reply:
`listResourceActions` throws an `InterruptedException` so this is needed. In other cases, we either aren't interacting with the `SamService` directly (although the underlying Sam call does have the `Rethrow.onInterrupted` handling) or the `SamService` call doesn't throw an `InterruptedException` (see `getSamUser`)